### ISSUE TITLE
Next.js throws `RangeError: Maximum call stack size exceeded` when running `yarn build`

### ISSUE THREAD

**Issue by danieltroger** [vercel/next.js#58494]
### Link to the code that reproduces this issue

https://github.com/danieltroger/react-server-components-stuff/tree/dt/next_infinite_loop

### To Reproduce

1. `git clone git@github.com:danieltroger/react-server-components-stuff.git`
2. `cd react-server-components-stuff`
3. `git checkout `dt/next_infinite_loop`
4. `yarn build`

### Current vs. Expected behavior

<details><summary>Current behavior</summary>

```
RangeError: Maximum call stack size exceeded
    at nO (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:60707)
    at nR (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:45694)
    at nT (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46289)
    at nO (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:62795)
    at nR (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:45694)
    at nT (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46289)
    at nO (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:62795)
    at nR (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:45694)
    at nT (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:46289)
    at nO (/private/tmp/nextjs/nextjs-app-router-examples/.yarn/__virtual__/next-virtual-d2b131c8dd/5/Users/daniel/.yarn/berry/cache/next-npm-14.0.3-canary.9-630a1159b2-10c0.zip/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:62795)
```
</details>

Expected behavior:
```
Error: Component nesting too deep, might exceed maximum call stack size soon. This usually happens if you have a component that recurses infinitely. Top of component stack:
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
SearchPage: app/components/SearchPage.tsx:4
```

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:43 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T6000
Binaries:
  Node: 20.5.0
  npm: 9.8.0
  Yarn: 4.0.1
  pnpm: 8.7.5
Relevant Packages:
  next: 14.0.3-canary.9
  eslint-config-next: 14.0.2
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

I wasted an embarrassingly long time on stepping through react internals for this (the function in the stack trace of the error that infinite loops' source code is here https://github.com/facebook/react/blob/593ecee66a609d4a4c2b36b39b1e5e23b2456dd1/packages/react-server/src/ReactFizzServer.js#L2168) when it's actually my fault. The reason this happened is because I forgot an import here: https://github.com/danieltroger/react-server-components-stuff/blob/dt/next_infinite_loop/app/components/SearchPage.tsx (it should be `import { SearchPage } from "@depict-ai/react-ui"`)

BUT the reason I'm reporting this as a bug is because my feeling is that next/react has the philosophy of displaying user friendly error messages when doing something wrong (I partly blame that I introduced this on running into `5:16  Error: Component definition is missing display name  react/display-name` so I called the function something without thinking about it). And I think this error message is the absolute opposite of user friendly. It leaves one two options: debugging react, which leads to nothing because the variables in scope where it infinite loops aren't helpful or just changing random things in ones app until one discovers the infinite loop.

**Comment by Kayeddy** [comment]
It seems that newer versions of Next JS has an issue with passing some type of elements (props) between client and server components. Given the case you're using MongoDB, you'll have to make sure you're correctly de-structuring query results that contain **ObjectId** data types, this will cause many errors, including the one you're describing. There are two ways I recommend to solve the issue: 

1. Using JSON.stringify(data) to turn the data into a plain text, so that there aren't any issues when sending down props between server/client components.

2. Taking the time to specify what is the content of the result data from each database query, specially if you're using Typescript. For example, instead of sending the entire retrieved data of an author down to a component and then trying to access each piece of the author information from within the component receiving the data, de-structure the data object and send the specific author information that the component will need.

To summarize, try to clean up and organize the data you receive from the backend instead of sending it raw to your components/pages.

**Comment by shawnesquivel** [comment]
Was running into the same issues:
- able to run `npm run build` no issues
- pushing to Vercel => errors

Now successfully working with:
-     "next": "^14.1.0", (migrated from 14.0)
-     "sharp": "^0.33.1" (downgraded from 0.33.2)

Doesn't work if you don't have sharp installed for some reason, even though I found conflicting information from several Github issues.

**Comment by Kayeddy** [comment]
> @Kayeddy not sure how your comment relates to this issue? There's no mongodb or JSON.stringify here

My apologies for the oversight, I kind of confused the context of the issue with another issue I saw on Stack Overflow.

Kind regards.

**Comment by SujithPriyamRajan** [comment]
I'm getting a similar kind of error. Don't know what's the issue.
![Screenshot 2024-07-29 151848](https://github.com/user-attachments/assets/44a17a85-b407-4022-9071-f42dbf194b7c)

**Comment by danieltroger** [comment]
@SujithPriyamRajan you're probably calling a component recursively, maybe the error message I detailed under "Expected behavior" helps explaining it?

**Comment by SujithPriyamRajan** [comment]
> @SujithPriyamRajan you're probably calling a component recursively, maybe the error message I detailed under "Expected behavior" helps explaining it?

I’m curious about this issue. Everything was working fine when using the client-side API, but since I moved the API to the server side, I’ve been facing problems with downloading blobs and creating window.URL.createObjectURL(blob)

**"use client"**
```
 const response = await api.viewDownload.download(path, FileTypes.CSV, details?.applicationName);
                const blobType = Object.getOwnPropertyDescriptor(response.headers, 'content-type')?.value;
                const blob = new Blob([response.data], {
                    type: blobType
                });

                const url = window.URL.createObjectURL(blob);
```
**"use server"**
```
 const response = await axios({
            url: fetchUrl.startsWith(config.REACT_APP_API_ENDPOINT) ? fetchUrl : '',
            method: options?.method,
            headers: options?.method?.toLowerCase() === 'get' ? { Accept: 'application/json' } : headers,
            data: options?.body
        });

        return options?.responseType === ResponseTypes.BLOB ? response : response.data.detail;
```

I have tried with https://github.com/vercel/next.js/issues/58494#issuecomment-1942427098

My current version is **"next": "14.2.3"**

**Comment by Nahuel-Salvatierra** [comment]
did you resolve this issue?

**Comment by BackSpaceTech** [comment]
Same problem here on Mac. 
Aborting installation.
Unexpected error. Please report it as a bug:
 RangeError: Maximum call stack size exceeded
    at openSync (node:fs:562:5)
Manual install of NextJs works fine though. Create-next-app fails. Is this resolved??
